LogInfo (etc) specific log level actions print a UserWarning in frontends

### Generated by Generative AI

_No response_

### Operating System:

Ubuntu 24.04 Noble

### ROS version or commit hash:

rolling

### RMW implementation (if applicable):

_No response_

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

_No response_

### 'ros2 doctor --report' output

<details><summary>ros2 doctor --report</summary>

```console
<COPY OUTPUT HERE>
```
</details>


### Steps to reproduce issue

my_launch.yaml

```yaml
---
launch:
  - log:
      level: info
      message: log with explicit info level
  - log_info:
      message: log_info
  - log_warning:
      message: log_warning
  - log_error:
      message: error
```

gives

```
ros2 launch my__launch.yaml
[INFO] [launch]: All log files can be found below /home/emerson/.ros/log/2025-05-16-00-03-54-473180-9a0947f3fde2-12989
[INFO] [launch]: Default logging verbosity is set to INFO
/ws/build/launch/launch/frontend/expose.py:107: UserWarning: The action log now expects a log level. Either provide one or switch to using the log_info action
  ret = found_parse_method(entity, parser)
[INFO] [launch.user]: log with explicit info level
[INFO] [launch.user]: log_info
[WARNING] [launch.user]: log_warning
[ERROR] [launch.user]: error

```



### Expected behavior

No warning!

### Actual behavior

The functionality works correctly! Just, the warning is printed misleadingly.

### Additional information

Introduced in https://github.com/ros2/launch/pull/866

--- a/launch/launch/actions/log.py
+++ b/launch/launch/actions/log.py
@@ -93,7 +93,7 @@ class Log(LogInterface):
         cls,
         entity: Entity,
         parser: 'Parser'
-    ):
+    ) -> Tuple[Type['Log'], Dict[str, Any]]:
         """Parse `log` tag."""
         _, kwargs = super().parse(entity, parser)
         kwargs['msg'] = parser.parse_substitution(entity.get_attr('message'))

--- a/launch/launch/actions/log.py
+++ b/launch/launch/actions/log.py
@@ -119,7 +119,7 @@ class SharedLogSpecificParse(Action):
         cls,
         entity: Entity,
         parser: 'Parser'
-    ):
+    ) -> Tuple[Type['SharedLogSpecificParse'], Dict[str, Any]]:
         """Parse `log_*` tag."""
         _, kwargs = super().parse(entity, parser)
         kwargs['msg'] = parser.parse_substitution(entity.get_attr('message'))

--- a/launch/launch/actions/log.py
+++ b/launch/launch/actions/log.py
@@ -31,9 +31,7 @@ from ..substitution import Substitution
 from ..utilities import normalize_to_list_of_substitutions
 
 
-@expose_action('log')
-class Log(Action):
-    """Action that logs a message when executed."""
+class LogInterface(Action):
 
     def __init__(self, *, msg: SomeSubstitutionsType,
                  level: SomeSubstitutionsType, **kwargs):
@@ -43,29 +41,6 @@ class Log(Action):
         self.__msg = normalize_to_list_of_substitutions(msg)
         self.__level = normalize_to_list_of_substitutions(level)
         self.__logger = launch.logging.get_logger('launch.user')
-
-    @classmethod
-    def parse(
-        cls,
-        entity: Entity,
-        parser: 'Parser'
-    ):
-        """Parse `log` tag."""
-        _, kwargs = super().parse(entity, parser)
-        kwargs['msg'] = parser.parse_substitution(entity.get_attr('message'))
-
-        # Check if still using old log action
-        level = entity.get_attr('level', optional=True)
-        # TODO: Remove optional level for Release after L-turtle release
-        if level is None:
-            warnings.warn(
-                'The action log now expects a log level.'
-                ' Either provide one or switch to using the log_info action',
-                stacklevel=2)
-            level = 'INFO'
-
-        kwargs['level'] = parser.parse_substitution(level)
-        return cls, kwargs
 
     @property
     def msg(self) -> List[Substitution]:
@@ -104,41 +79,79 @@ class Log(Action):
         return None
 
 
+@expose_action('log')
+class Log(LogInterface):
+    """Action that logs a message when executed."""
+
+    @classmethod
+    def parse(
+        cls,
+        entity: Entity,
+        parser: 'Parser'
+    ):
+        """Parse `log` tag."""
+        _, kwargs = super().parse(entity, parser)
+        kwargs['msg'] = parser.parse_substitution(entity.get_attr('message'))
+
+        # Check if still using old log action
+        level = entity.get_attr('level', optional=True)
+        # TODO: Remove optional level for Release after L-turtle release
+        if level is None:
+            warnings.warn(
+                'The action log now expects a log level.'
+                ' Either provide one or switch to using the log_info action',
+                stacklevel=2)
+            level = 'INFO'
+
+        kwargs['level'] = parser.parse_substitution(level)
+        return cls, kwargs
+
+
+class SharedLogSpecificParse(Action):
+
+    @classmethod
+    def parse(
+        cls,
+        entity: Entity,
+        parser: 'Parser'
+    ):
+        """Parse `log_*` tag."""
+        _, kwargs = super().parse(entity, parser)
+        kwargs['msg'] = parser.parse_substitution(entity.get_attr('message'))
+        return cls, kwargs
+
+
 @expose_action('log_info')
-class LogInfo(Log):
+class LogInfo(SharedLogSpecificParse, LogInterface):
     """Action that logs a message with level INFO when executed."""
 
     def __init__(self, *, msg: SomeSubstitutionsType, **kwargs):
         """Create a LogInfo action."""
-        kwargs.pop('level', None)
         super().__init__(msg=msg, level='INFO', **kwargs)
 
 
 @expose_action('log_warning')
-class LogWarning(Log):
+class LogWarning(SharedLogSpecificParse, LogInterface):
     """Action that logs a message with level WARNING when executed."""
 
     def __init__(self, *, msg: SomeSubstitutionsType, **kwargs):
         """Create a LogWarning action."""
-        kwargs.pop('level', None)
         super().__init__(msg=msg, level='WARNING', **kwargs)
 
 
 @expose_action('log_debug')
-class LogDebug(Log):
+class LogDebug(SharedLogSpecificParse, LogInterface):
     """Action that logs a message with level DEBUG when executed."""
 
     def __init__(self, *, msg: SomeSubstitutionsType, **kwargs):
         """Create a LogDebug action."""
-        kwargs.pop('level', None)
         super().__init__(msg=msg, level='DEBUG', **kwargs)
 
 
 @expose_action('log_error')
-class LogError(Log):
+class LogError(SharedLogSpecificParse, LogInterface):
     """Action that logs a message with level ERROR when executed."""
 
     def __init__(self, *, msg: SomeSubstitutionsType, **kwargs):
         """Create a LogError action."""
-        kwargs.pop('level', None)
         super().__init__(msg=msg, level='ERROR', **kwargs)

--- a/launch_xml/test/launch_xml/test_log.py
+++ b/launch_xml/test/launch_xml/test_log.py
@@ -44,8 +44,10 @@ def test_log():
         """
     xml_file = textwrap.dedent(xml_file)
     root_entity, parser = load_no_extensions(io.StringIO(xml_file))
-    with pytest.warns(Warning):
+    with pytest.warns(Warning) as record:
         launch_description = parser.parse_description(root_entity)
+
+    assert len(record) == 1
 
     log = launch_description.entities[0]
     assert isinstance(log, Log)

--- a/launch_yaml/test/launch_yaml/test_log.py
+++ b/launch_yaml/test/launch_yaml/test_log.py
@@ -50,8 +50,10 @@ def test_log():
         """
     yaml_file = textwrap.dedent(yaml_file)
     root_entity, parser = load_no_extensions(io.StringIO(yaml_file))
-    with pytest.warns(Warning):
+    with pytest.warns(Warning) as record:
         launch_description = parser.parse_description(root_entity)
+
+    assert len(record) == 1
 
     log = launch_description.entities[0]
     assert isinstance(log, Log)

